zipkin-lens/src/components/TracePage/SpanDetail/SpanAnnotation.jsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2015-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import PropTypes from 'prop-types';
+import React from 'react';
+import { withStyles } from '@material-ui/styles';
+import Box from '@material-ui/core/Box';
+import Table from '@material-ui/core/Table';
+import TableBody from '@material-ui/core/TableBody';
+import TableCell from '@material-ui/core/TableCell';
+import TableRow from '@material-ui/core/TableRow';
+import Paper from '@material-ui/core/Paper';
+
+import { spanAnnotationPropTypes } from '../../../prop-types';
+import { formatTimestampMillis } from '../../../util/timestamp';
+
+const propTypes = {
+  annotation: spanAnnotationPropTypes.isRequired,
+  classes: PropTypes.shape({}).isRequired,
+};
+
+const style = theme => ({
+  cell: {
+    paddingTop: theme.spacing(1),
+    paddingBottom: theme.spacing(1),
+  },
+});
+
+const SpanAnnotation = ({ annotation, classes }) => (
+  <Box>
+    <Box fontSize="1.1rem" mb={0.5}>
+      {annotation.value}
+    </Box>
+    <Paper>
+      <Table>
+        <TableBody data-testid="span-annotation--table-body">
+          {
+            [
+              { label: 'Start Time', value: formatTimestampMillis(annotation.timestamp) },
+              { label: 'Relative Time', value: annotation.relativeTime },
+              { label: 'Address', value: annotation.endpoint },
+            ].map(e => (
+              <TableRow key={e.label}>
+                <TableCell className={classes.cell} data-testid="span-annotation--label">
+                  {e.label}
+                </TableCell>
+                <TableCell className={classes.cell} data-testid="span-annotation--value">
+                  {e.value}
+                </TableCell>
+              </TableRow>
+            ))
+          }
+        </TableBody>
+      </Table>
+    </Paper>
+  </Box>
+);
+
+SpanAnnotation.propTypes = propTypes;
+
+export default withStyles(style)(SpanAnnotation);
```

Comment:
I want to think a little more about whether to use `React.memo` for all the components.
But using `React.memo` on the trace timeline clearly improved rendering performance.

**Before**

![before-optimization](https://user-images.githubusercontent.com/19551419/65944886-a4dc9b00-e46d-11e9-99d5-1ea8935418f1.gif)

**After**

![after-optimization](https://user-images.githubusercontent.com/19551419/65944905-b02fc680-e46d-11e9-8d63-eeb30f60f69b.gif)

Thank you for your advice!

Reply:
That's a nice improvement! I'm guessing there might be other good places like each individual trace (make sure not to rerender when expanding a trace) - but often it's just simplest to reason about to use `memo` as much as possible since

- We have so many components anyways, adding the HoC isn't a significant impact
- Either a component rerenders, meaning the computation of memo was a bit of a waste, or it doesn't meaning it was worth it. But even for very simple components the rerender (virtual DOM operations) should always be much more expensive then computing the memo.

But will leave it to you how much you add it :)